app/workers/messaging.php

```diff
@@ -20,59 +33,197 @@
-        );
+        $targets = $this->dbForProject->find('targets', [Query::equal('$id', $recipientsId)]);
+        $recipients = \array_merge($recipients, $targets);
+        $recipients = \array_filter($recipients, fn (Document $recipient) => $recipient->getAttribute('providerId') === $providerRecord->getId());
+
+        $identifiers = \array_map(function (Document $recipient) {
+            return $recipient->getAttribute('identifier');
+        }, $recipients);
+
+        $maxBatchSize = $provider->getMaxMessagesPerRequest();
+        $batches = \array_chunk($identifiers, $maxBatchSize);
 
-        try {
-            $this->sms->send($message);
-        } catch (\Exception $error) {
-            throw new Exception('Error sending message: ' . $error->getMessage(), 500);
+        $results = batch(\array_map(function ($batch) use ($messageRecord, $providerRecord, $provider) {
+            return function () use ($batch, $messageRecord, $providerRecord, $provider) {
+                $deliveredTo = 0;
+                $deliveryErrors = [];
+                $messageData = clone $messageRecord;
+                $messageData->setAttribute('to', $batch);
+                $message = match ($providerRecord->getAttribute('type')) {
+                    'sms' => $this->buildSMSMessage($messageData, $providerRecord),
+                    'push' => $this->buildPushMessage($messageData),
+                    'email' => $this->buildEmailMessage($messageData, $providerRecord),
+                    default => throw new Exception(Exception::PROVIDER_INCORRECT_TYPE)
+                };
+                try {
+                    $provider->send($message);
+                    $deliveredTo += \count($batch);
+                } catch (\Exception $e) {
+                    foreach ($batch as $identifier) {
+                        $deliveryErrors[] = 'Failed to send message to target' . $identifier . ': ' . $e->getMessage();
+                    }
+                } finally {
+                    return [
+                        'deliveredTo' => $deliveredTo,
+                        'deliveryErrors' => $deliveryErrors,
+                    ];
+                }
+            };
+        }, $batches));
+
+        $deliveredTo = 0;
+        $deliveryErrors = [];
+        foreach ($results as $result) {
+            $deliveredTo += $result['deliveredTo'];
+            $deliveryErrors = \array_merge($deliveryErrors, $result['deliveryErrors']);
+        }
+        $messageRecord->setAttribute('deliveryErrors', $deliveryErrors);
+
+        if (\count($messageRecord->getAttribute('deliveryErrors')) > 0) {
+            $messageRecord->setAttribute('status', 'failed');
+        } else {
+            $messageRecord->setAttribute('status', 'sent');
         }
+        $messageRecord->setAttribute('to', $recipientsId);
+        $messageRecord->setAttribute('deliveredTo', $deliveredTo);
+        $messageRecord->setAttribute('deliveryTime', DateTime::now());
```

Comment:
We shouldn't update this attribute, this should just be the time the developer wants the message to be sent.

But this info would be useful, maybe we should add an `deliveredAt` attribute that contains the time the message was actually delivered?

Reply:
done this